python/ray/rllib/ddpg/ddpg_evaluator.py

```diff
@@ -0,0 +1,213 @@
+            new_priorities = (
+                np.abs(td_errors) + self.config["prioritized_replay_eps"])
+            batch.data["weights"] = new_priorities
+
+        return batch
+
+    def compute_gradients(self, samples):
+        td_err, grads = self.ddpg_graph.compute_gradients(
+            self.sess, samples["obs"], samples["actions"], samples["rewards"],
+            samples["new_obs"], samples["dones"], samples["weights"])
+        return grads, {"td_error": td_err}
+
+    def apply_gradients(self, grads):
+        self.ddpg_graph.apply_gradients(self.sess, grads)
+
+    def compute_apply(self, samples):
+        td_error = self.ddpg_graph.compute_apply(
+            self.sess, samples["obs"], samples["actions"], samples["rewards"],
+            samples["new_obs"], samples["dones"], samples["weights"])
+        return {"td_error": td_error}
+
+    def get_weights(self):
+        return self.variables.get_weights()
+
+    def set_weights(self, weights):
+        self.variables.set_weights(weights)
+
+    def _step(self, global_timestep):
+        """Takes a single step, and returns the result of the step."""
+        action = self.ddpg_graph.act(
+            self.sess, np.array(self.obs)[None],
+            self.exploration.value(global_timestep))[0]
+        new_obs, rew, done, _ = self.env.step(action)
+        ret = (self.obs, action, rew, new_obs, float(done))
+        self.obs = new_obs
+        self.episode_rewards[-1] += rew
+        self.episode_lengths[-1] += 1
+        if done:
+            self.obs = self.env.reset()
+            self.episode_rewards.append(0.0)
+            self.episode_lengths.append(0.0)
+            # reset UO noise for each episode
+            self.ddpg_graph.reset_noise(self.sess)
+
+        self.local_timestep += 1
+        return ret
+
+    def stats(self):
+        n = self.config["smoothing_num_episodes"] + 1
+        mean_100ep_reward = round(np.mean(self.episode_rewards[-n:-1]), 5)
+        mean_100ep_length = round(np.mean(self.episode_lengths[-n:-1]), 5)
+        exploration = self.exploration.value(self.global_timestep)
+        return {
+            "mean_100ep_reward": mean_100ep_reward,
+            "mean_100ep_length": mean_100ep_length,
+            "num_episodes": len(self.episode_rewards),
+            "exploration": exploration,
+            "local_timestep": self.local_timestep,
+        }
+
+    def save(self):
+        return [
+            self.exploration,
+            self.episode_rewards,
+            self.episode_lengths,
+            self.saved_mean_reward,
+            self.obs,
+            self.global_timestep,
+            self.local_timestep]
+
+    def restore(self, data):
+        self.exploration = data[0]
+        self.episode_rewards = data[1]
+        self.episode_lengths = data[2]
+        self.saved_mean_reward = data[3]
+        self.obs = data[4]
+        self.global_timestep = data[5]
+        self.local_timestep = data[6]
```

Comment:
I'm a little concerned this file has too much in common with `dqn_evaluator.py`. However, it's not clear if coupling DQN and DDPG would be a good idea either. @richardliaw any thoughts?

Reply:
You are right. Both of them are in a Q-learning behavior. The differences can be expressed by DQNGraph and DDPGGraph. We can consider distilling a parent class like QAgent/QEvaluator later.